pubnub-core/src/subscription/subscribe_loop.rs

```diff
@@ -0,0 +1,282 @@
+use super::subscribe_request;
+use crate::message::Timetoken;
+use crate::transport::Transport;
+use futures_channel::{mpsc, oneshot};
+use futures_util::future::{select, Either, FutureExt};
+use futures_util::sink::SinkExt;
+use futures_util::stream::StreamExt;
+use log::{debug, error};
+
+pub(crate) use super::channel::{Rx as ChannelRx, Tx as ChannelTx};
+pub(crate) use super::registry::ID as SubscriptionID;
+
+pub(crate) type Registry = GenericRegistry<ChannelTx>;
+
+pub(crate) type ReadyTx = oneshot::Sender<()>;
+#[allow(dead_code)]
+pub(crate) type ReadyRx = oneshot::Receiver<()>;
+
+pub(crate) type ExitTx = mpsc::Sender<()>;
+#[allow(dead_code)]
+pub(crate) type ExitRx = mpsc::Receiver<()>;
+
+pub(crate) type ControlTx = mpsc::Sender<ControlCommand>;
+pub(crate) type ControlRx = mpsc::Receiver<ControlCommand>;
+
+pub(crate) type SubscriptionIdTx = oneshot::Sender<SubscriptionID>;
+#[allow(dead_code)]
+pub(crate) type SubscriptionIdRx = oneshot::Receiver<SubscriptionID>;
+
+/// Commands we pass via the control pipe.
+#[derive(Debug)]
+pub(crate) enum ControlCommand {
+    /// A stream for a channel or channel group is being dropped.
+    ///
+    /// Only sent from `Subscription` to `SubscribeLoop`.
+    Drop(SubscriptionID, ListenerType),
+
+    /// A stream for a channel or channel group is being created.
+    ///
+    /// Only sent from `PubNub` to `SubscribeLoop`.
+    Add(ListenerType, ChannelTx, SubscriptionIdTx),
+}
+
+/// # Type of listener (a channel or a channel group)
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ListenerType {
+    Channel(String), // Channel name
+    _Group(String),  // Channel Group name
+}
+
+#[derive(Debug)]
+pub(crate) struct SubscribeLoopParams<TTransport> {
+    pub control_rx: ControlRx,
+    pub ready_tx: Option<ReadyTx>,
+    pub exit_tx: Option<ExitTx>,
+
+    pub transport: TTransport,
+
+    pub origin: String,
+    pub agent: String,
+    pub subscribe_key: String,
+
+    pub channels: Registry,
+    pub groups: Registry,
+}
+
+/// Implements the subscribe loop, which efficiently polls for new messages.
+pub(crate) async fn subscribe_loop<TTransport>(params: SubscribeLoopParams<TTransport>)
+where
+    TTransport: Transport,
+    <TTransport as Transport>::Error: 'static,
+{
+    debug!("Starting subscribe loop");
+
+    #[allow(clippy::unneeded_field_pattern)]
+    let SubscribeLoopParams {
+        mut control_rx,
+        mut ready_tx,
+        mut exit_tx,
+
+        transport,
+
+        origin,
+        agent: _, // TODO: use it.
+        subscribe_key,
+
+        mut channels,
+        mut groups,
+    } = params;
+
+    let mut encoded_channels = EncodedChannelsList::from(&channels);
+    let mut encoded_groups = EncodedChannelsList::from(&groups);
+
+    let mut timetoken = Timetoken::default();
+
+    loop {
+        // TODO: do not clone cached values cause it defeats the purpose of the
+        // cache. Do we need this cache though? There are no benchmarks...
```

Comment:
This clone can be removed if `subscribe_request::subscribe_request` doesn't borrow the string for the lifetime of its return value.

As for whether the cache is necessary, my gut instinct tells me that any serde (percent-encoding, JSON, etc) is a huge time sink. And I'm happy to spend a little memory to free the CPU for more important tasks.

Reply:
> This clone can be removed if `subscribe_request::subscribe_request` doesn't borrow the string for the lifetime of its return value.

Right! I'll separate the lifetimes, this should solve this issue.

> As for whether the cache is necessary, my gut instinct tells me that any serde (percent-encoding, JSON, etc) is a huge time sink. And I'm happy to spend a little memory to free the CPU for more important tasks.

I absolutely agree with you about the gut instinct, I have this too. However, it looks like yet another premature optimization case. I don't mind it now, but I think this part can safely be removed if it causes flexibility issues when we switch from using `url`s in `Transport`.

Reply:
I really like where this is going. It will DRY up a great deal of the data interfaces in here.

Reply:
Just another gut feeling here, but it seems like the cache would just have to move closer to the `Transport` if the types were improved. Not necessarily that removing it entirely would be ideal.

Reply:
Oh, yeah, I think that'd be the case! Good point!

Reply:
I updated the TODO to capture the decision to move the cache to the transport.